Add filter for optional blocking of file uploads in upload_handler()

mreishus opened the pull request:
Related to #https://github.com/Automattic/wp-calypso/issues/51594

#### Changes proposed in this Pull Request:
* Add a new filter which can be used to block uploads processed by remote_request_handlers() -> upload_handler() 
  *  For example, a WP.com user using calypso to upload to a Jetpack or Atomic site

An example use of the filter is 832-gh-Automattic/wpcomsh . This prevents WoA users from using `https://wordpress.com/media/<site>` to upload files to their site, if that upload would take them over disk quota. Currently, this upload path goes through  `(wpcom) upload_jetpack_files() -> Jetpack_Client::upload_files() --> curl` `--http-->` `(jetpack) do_action 'wp_ajax_nopriv_jetpack_upload_file' -> remote_request_handlers() -> upload_handler()`. No file upload size checking is done in this path, as far as I can tell.  Adding the filter allows WoA to do their own checking of file sizes.

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
* No

#### Testing instructions:
* Create a Jetpack site running this branch
* Connect to WP.com
* Add this code somewhere
```
function test_jetpack_upload_handler_can_upload( $value, $files ) {
        $upload_size = 0;
        if ( ! empty( $files['media']['size'] ) ) {
                $upload_size = array_sum( $files['media']['size'] );
                if ( $upload_size > 50000 ) {
                        return new WP_Error( 'insufficient_space_available', 'Uploaded file is too large.' );
                }
        }

        return $value;
}
add_filter( 'jetpack_upload_handler_can_upload', 'test_jetpack_upload_handler_can_upload', 10, 2 );
```
* Visit `https://wordpress.com/media/<sitename>`
* You should be able to upload a very small image (below 50kb), but a large image (above 50kb) should return an error and not write a file to disk

![2022-03-03_16-41](https://user-images.githubusercontent.com/937354/156665106-7148a724-c1df-4940-9fd4-f141ef4b1078.png)

mmtr commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -3913,6 +3913,11 @@ public function upload_handler( $update_media_item = false ) {
 			return new WP_Error( 'unknown_token', 'Unknown Jetpack token', 403 );
 		}
 
+		$upload_blocked = apply_filters( 'jetpack_upload_handler_block', false, $_FILES );
```
Can you add a docstring for this filter? Something similar to this format would be great:

https://github.com/Automattic/jetpack/blob/05f9ad44f1725e736d38bdf080d02b641d850b8f/projects/plugins/jetpack/class.jetpack.php#L4870-L4878

mmtr commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -3913,6 +3913,11 @@ public function upload_handler( $update_media_item = false ) {
 		}
 
+		$upload_blocked = apply_filters( 'jetpack_upload_handler_block', false, $_FILES );
+		if ( is_wp_error( $upload_blocked ) ) {
```
Nitpick (non-blocking): `jetpack_upload_handler_block` can set the wrong expectation that returning a `true` value  will also block the file upload, since the default value of the filter is `false`.

Do you think that using something more _WordPress-y_,  like `jetpack_upload_handler_can_upload` where `true` (default) proceeds with the upload and `false` or `WP_Error` blocks it, could help setting better expectations?

mmtr reviewed (commented):
This works great. I left a couple of minor comments.

coder-karen reviewed (changes requested):
Since this will end up on it's own hook page on developer.jetpack.com, I'd suggest changing `jetpack` to `Jetpack` on line 3917.

coder-karen commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -3913,6 +3913,23 @@ public function upload_handler( $update_media_item = false ) {
 		}
 
+		/**
+		 * Optionally block uploads processed through jetpack's upload_handler().
```
```suggestion
		 * Optionally block uploads processed through Jetpack's upload_handler().
```

Since this will end up on it's own hook page on developer.jetpack.com, I'd suggest changing jetpack to Jetpack here.

mreishus replied:
Fixed in 7c6ac6e381

coder-karen commented:
I tried to test the PR out but was unable to get it working (I'm possibly doing something wrong at my end - have been adding the snippet via the Code Snippets plugin on two test Jetpack sites that have the updated `class.jetpack.php` file in it and attempting to upload large-ish media via the mentioned link), so another Crew member will still need to review :)

mreishus commented:
Rebased + Fixed capitalization in comment

mreishus commented:
> I tried to test the PR out but was unable to get it working (I'm possibly doing something wrong at my end - have been adding the snippet via the Code Snippets plugin on two test Jetpack sites that have the updated `class.jetpack.php` file in it and attempting to upload large-ish media via the mentioned link), so another Crew member will still need to review :)

Just checked; the snippet I had in the original PR description referred to the old name of the hook. During this PR, we changed the name of the hook but I never updated the snippet in the description. This might be what you're seeing. I've updated the snippet to use the new name of the hook (`jetpack_upload_handler_can_upload`):

```
function test_jetpack_upload_handler_can_upload( $value, $files ) {
        $upload_size = 0;
        if ( ! empty( $files['media']['size'] ) ) {
                $upload_size = array_sum( $files['media']['size'] );
                if ( $upload_size > 50000 ) {
                        return new WP_Error( 'insufficient_space_available', 'Uploaded file is too large.' );
                }
        }

        return $value;
}
add_filter( 'jetpack_upload_handler_can_upload', 'test_jetpack_upload_handler_can_upload', 10, 2 );
```

I was able to test this on a jetpack docker site.  I don't know how to install plugins on jetpack docker sites, so I added the snippet above the the bottom of `wp-content/plugins/hello.php`, then activated the Hello Dolly plugin. On jetpack docker, I also had to chmod the uploads directory so that calypso could upload into it:
```
$ docker exec -it 087beb9bfc7e /bin/bash 
root@087beb9bfc7e:/var/www/html# cd wp-content
root@087beb9bfc7e:/var/www/html/wp-content# chmod 777 .
```

Then I was able to upload through calypso and have a small file (<50KB) make it through, but a larger file (>50KB) was blocked)
![2022-03-07_13-50](https://user-images.githubusercontent.com/937354/157108061-58fda5e0-417c-4526-a5e8-828488431a8c.png)

jeherve commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -3913,6 +3913,23 @@ public function upload_handler( $update_media_item = false ) {
+		 * Optionally block uploads processed through Jetpack's upload_handler().
+		 * The filter may return false or WP_Error to block this particular upload.
+		 *
+		 * @since 10.8
```
```suggestion
		 * @since $$next-version$$
```

jeherve commented on projects/plugins/jetpack/class.jetpack.php:
```diff
@@ -3913,6 +3913,23 @@ public function upload_handler( $update_media_item = false ) {
+			if ( is_wp_error( $can_upload ) ) {
+				return $can_upload;
+			}
+			return new WP_Error( 'handler_cannot_upload', 'The upload handler cannot upload files', 400 );
```
```suggestion
			return new WP_Error( 'handler_cannot_upload', __( 'The upload handler cannot upload files', 'jetpack' ), 400 );
```

jeherve reviewed (approved):
This tests well, it should be good to merge. 👍
